### PR TITLE
Feature: DynamoDB Lock

### PR DESCRIPTION
**Pull request by ina-stoyanova**
Adding a DynamoDB lock feature based on the code from [prototypes/release-notes-drafter](https://github.com/gruntwork-io/prototypes/blob/65c81b6/release-notes-drafter/aws.go).

Related: https://github.com/gruntwork-io/go-commons/issues/44

### CODE REVIEW

File: go.sum
Comment (marinalimeira) [inline go.sum:1]: Merging with master had conflicts at `go.mod` and `go.sum`,  running `go mod tidy` resulted in these changes

File: go.mod
Comment (marinalimeira) [inline go.mod:6]: Merging with master had conflicts at `go.mod` and `go.sum`,  running `go mod tidy` resulted in these changes

Comment (ina-stoyanova) [reply go.mod:6]: Thanks! Is there any action needed?

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:29]: I think these days, to work with AWS SSO, we do:

```go
sessionOptions := session.Options{
    SharedConfigState: session.SharedConfigEnable, 
    Config: aws.NewConfig().WithRegion(awsRegion),
}
sess, err := session.NewSessionWithOptions(sessionOptions)
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:100]:
```suggestion
		"Attempting to acquire lock %s in table %s in region %s, retrying on failure for up to %d times with %s between retries",
```

Comment (brikis98) [reply lock/lock.go:100]: NIT: the `retry.DoWithRetry` method called below might already write such a log message, so this might be redundant.

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:91]: NIT: Hm, I'm not sure if blocking is quite the right term here. Perhaps `AcquireLockWithRetries` would be clearer?

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:18]: Who creates the table?

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:16]: Use unique ID to ensure other tests running concurrently aren't competing for the same lock? Same for the table name too.

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:29]: Perhaps `require.NoError` so the test doesn't keep running if this first step fails?

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:36]: Use `require.Error` and check the error message is what we expect

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:44]: NIT: extract into `GetLockStatus` (or similar) method? Might be a useful part of the public API.

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:53]: This needs more context: namely, that we are using a DynamoDB table for locking, and we are using it in such a way that this will work as a distributed lock (as opposed to, say, a local lock for multi threading).

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go:100]: yeah, makes sense.

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go:91]: good shout, done.

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:18]: Some thoughts from discussing:
- so far it's been done manually 
- we want to create a table in a consistent way, only in one central location, one region

the plan is: 
- add a new interface to allow creation of a dynamoDB table with these params.

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go]: good shout - added a new method for that!

**Comment by ina-stoyanova** [comment]
I've gone ahead and addressed all of the above OR feedback. Another set of eyes will be very much appreciated :)

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:22]:
```suggestion
const (
	// Terraform requires the DynamoDB table to have a primary key with this name
	attributeLockId = "LockID"

	// Default is to retry for up to 5 minutes
	maxRetriesWaitingForTableToBeActive = 30
	sleepBetweenTableStatusChecks       = 10 * time.Second

	dynamoDbPayPerRequestBillingMode = "PAY_PER_REQUEST"
)
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:81]:
```suggestion
// AcquireLock will attempt to acquire a lock in DynamoDB table while taking the configuration options into account.
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:82]:
```suggestion
// We are using DynamoDB to create a table to help us track the lock status of different resources.
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:141]:
```suggestion
// acquireLockWithRetries will attempt to acquire the lock defined by the provided lock string in the configured lock table
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:177]:
```suggestion
		fmt.Sprintf("Trying to release DynamoDB lock %s at table %s\n", options.LockString, options.LockTable),
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:223]:
```suggestion
// createLockTableIfNecessary will create the lock table in DynamoDB if it doesn't already exist
```

File: lock/lock.go
Comment (bwhaley) [inline lock/lock.go:296]:
```suggestion
// lockTableExistsAndIsActive will return true if the lock table exists in DynamoDB and is in "active" state
```

File: lock/lock_test.go
Comment (bwhaley) [inline lock/lock_test.go:38]: This is a good place to use [`assert.Error`](https://pkg.go.dev/github.com/stretchr/testify/assert#Error) to make the code slightly more legible. Note that the trailing newline [is automatically appended](https://github.com/stretchr/testify/blob/acba37e5db06f0093b465a7d47822bf13644b66c/assert/assertions.go#L262) and the `.` is kind of optional IMO for test error messages.
```suggestion
	assert.Error(t, err, "Acquiring of second lock succeeded, but it was expected to fail")
```

**Review by bwhaley — DISMISSED** [review]
Left a few nits for you to consider.

**Comment by ina-stoyanova** [comment]
Thanks for the review and approval @bwhaley :)

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:31]: NIT: add `//` comments next to each field explaining what it is.

NIT: put these in a more "story like" order. E.g., `Logger` is least important, so put it at the end. `LockTable` and `AwsRegion` should prob be at the top; then `LockString`; then the two retry params; etc.

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:98]: NIT: A way to reduce the code for this pattern by one line:

```suggestion
		if err := createLockTableIfNecessary(options, client); err != nil {
			return errors.WithStackTrace(err)
		}
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:117]: Remove and use `client` instead

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:137]: NIT: 

```suggestion
	options.Logger.Infof("Acquired lock '%s' in table %s\n", options.LockString, options.LockTable)
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:146]:
```suggestion
		fmt.Sprintf("Trying to acquire DynamoDB lock %s in table %s\n", options.LockString, options.LockTable),
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:170]: NIT:

```suggestion
	if !tableExists {
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:175]: NIT: for consistency with `acquireLockWithRetries`, either extract this into `releaseLockWithRetries`, or, alternatively, remove `acquireLockWithRetries`, and inline that logic within `AcquireLock`.

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:199]: Remove and use `client`

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:203]: Replace `LockID` with constant already defined above

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:219]: NIT:

```suggestion
	options.Logger.Infof("Released lock '%s' in table %s\n", options.LockString, options.LockTable)
```

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:328]: Wrong error text

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:13]: You want to do this within the TestXXX method below. That way, if there are multiple test methods, each one creates its own `Options`, with different (unique) lock tables / strings.

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:16]: Use `random.UniqueId()` in the table name too. This ensures (a) that the table creation logic is put through its paces and (b) that if two tests run concurrently, one of them doesn't delete the table the other is using.

File: lock/lock_test.go
Comment (brikis98) [reply lock/lock_test.go:13]: Also, when instantiating structs, its best to pass the field names explicitly, rather than relying on ordering, which is error prone, and hard to read. E.g.,

```go
options := Options {
  Logger: logging.getLogger("test"),
  LockTable: xxx,
  LockString: yyy,
  // etc
}
```

**Comment by ina-stoyanova** [comment]
Thanks, Jim! That's all really great feedback from you and Ben! I'll get to this today 🙂

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:38]: Oh, that's actually how it was before. I believe there was a brief comment from Jim here that it might be easier to compare the error message and use the `require`. I've looked into what happens under the hood for the 2 methods, and the outcome for this case seems to be the same, although under the hood they'll call into different underlying functions. 

I'm happy to proceed with it being assert

Comment (ina-stoyanova) [reply lock/lock_test.go:38]: @brikis98 not super serious, but curious to see what you make of this - is there a compelling reason why you thought the `require` is a better choice here?

Comment (brikis98) [reply lock/lock_test.go:38]: There's no need for the `if err == nil` check. It's _just_ the one-liner with `require.xxx` or `assert.xxx`. Whether you use `require` or `assert` depends on whether the test should keep running or not. But the point is to replace a 3-line `if err = nil { ... }` block with a one-liner (which also has better error reporting).

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go]: I added the comments - I'll have to re-read them once again though to see if they make sense

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go:130]: ah great spot! Totally missed it

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go:216]: also a good spot! thanks 😄

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:13]: that's all very good points - thanks a lot Jim!

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:16]: 👍  I think this also means then the test will need to delete the DynamoDB table itself to clean up.

**Comment by ina-stoyanova** [comment]
Next todo: 
- fix build
- add delete dynamoDB table so test can cleanup after itself
- review comments if they make sense
- run test & confirm intermittent errors seen are expected
- use functionality from cis reop branch & see if the test there still passes

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:38]: Ahh! Of course :) Noted 👍

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go]: oops - fixed now. I'll add some more details to it too

**Comment by brikis98** [comment]
Is this ready for re-review?

**Comment by ina-stoyanova** [comment]
Just coming back to this now! I'll ping you again! I'm following up on this:

- [ ] add delete dynamoDB table so test can cleanup after itself
- [ ] review comments if they make sense
- [ ] run test & confirm intermittent errors seen are expected
- [ ] use functionality from cis reop branch & see if the test there still passes
- [ ] make sure error handling and logging makes sense and does not add extra code for no reason

**Comment by marinalimeira** [comment]
Updates on TicketJam:
- remove cleanup, it causes concurrency errors if a test is waiting to lock while another test finished running and deleted the table.
- update the test table to always use the same table, but different lock id strings.

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:188]: I may be missing something, but why do we need to do the release lock in a retry loop? It makes sense to _acquire_ a lock in a retry loop, as you are waiting for someone else to release it, but shouldn't _release_ be a one-time action? Either you can release it, or something is wrong?

File: lock/lock_test.go
Comment (brikis98) [inline lock/lock_test.go:35]: I thought `assert.Error`, as the 3rd argument, allowed you to check the contents of the error? In this case, we'd want to be sure that the error we got was that the lock was already acquired, and not some other error (e.g., failure to talk to DynamoDB).

File: lock/lock.go
Comment (ina-stoyanova) [reply lock/lock.go:188]: Hm, you might be right actually! I thought it would make sense to retry it in case of an error, but we _should probably_ just throw the error.

File: lock/lock_test.go
Comment (ina-stoyanova) [reply lock/lock_test.go:35]: Yeah, that's correct. This is a leftover actually from the first time around. I'll change it to
```suggestion
		assert.Error(t, err,
		fmt.Sprintf("Error acquiring lock %s in table %s in region %s (already locked?): %s\n",
			options.LockString,
			options.LockTable,
			options.AwsRegion,
			err))
```

**Comment by ina-stoyanova** [comment]
I believe this is ready for hopefully a final set of eyes @brikis98 if you've got the chance :)

File: lock/lock.go
Comment (brikis98) [inline lock/lock.go:39]: Sorry, one more tiny thing: the `ReleaseLock` thing no longer uses retries, so these comments prob need to be updated.

Comment (ina-stoyanova) [reply lock/lock.go:39]: Very good catch - I totally missed it! Thank you

File: lock/lock.go
Comment (ina-stoyanova) [inline lock/lock.go:37]:
```suggestion
	// The value for how many times AcquireLock will retry to acquire the lock
```

File: lock/lock.go
Comment (ina-stoyanova) [inline lock/lock.go:39]:
```suggestion
	// The value for how long `AcquireLock` will sleep for between retries to get the lock
```

File: lock/lock.go
Comment (ina-stoyanova) [inline lock/lock.go:264]:
```suggestion
// waitForTableToBeActive will wait for the given DynamoDB table to be in the "active" state. If it's not in "active" state, this function will sleep for the
// specified amount of time, and try again, up to a maximum of maxRetries retries. Note this is different from the MaxRetires value for how many times to retry when acquiring the lock.
```

File: lock/lock.go
Comment (ina-stoyanova) [inline lock/lock.go:298]:
```suggestion
// GetLockStatus attempts to acquire the lock and check if the expected item is there
// If there's the expected Item with the correct `LockString` value - then the status is `locked`, if the item is not there - then the status is `not locked`
```

**Comment by ina-stoyanova** [comment]
So, I've update a few more comments actually - hope they make sense now and are useful! Thanks for all the reviews here :)

**Review by brikis98 — APPROVED** [review]
LGTM!

**Comment by ina-stoyanova** [comment]
Thanks for the review, Jim! I'm merging and releasing this!

**Pull request merged**
